### PR TITLE
feat(schema-form): add checkboxes widget

### PR DESCRIPTION
**Pull request by djcouto**
Screenshot 2022-11-10 at 08 56 51

### CODE REVIEW

**Review by GoncaloCamaz — APPROVED** [review]

**Review by nelsonmestevao — APPROVED** [review]

**Review by Insatisfeito — APPROVED** [review]

File: packages/react-ui-components/src/components/SchemaForm/Widgets/CheckboxesWidget/utils.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import { ICheckboxConfig } from './types';
+
+export const buildRadioButtons = <T extends { label: string; value: string }>(options: T[], enumDisabled: string[], { multiple, allButton }: ICheckboxConfig): IRadioButton[] =>
+	options.reduce<IRadioButton[]>(
+		(accumulator, { label, value }) => [
+			...accumulator,
+			{
+				label,
+				value,
+				disabled: enumDisabled.indexOf(value) !== -1
```
Comment (ricardogoncalves89) [inline packages/react-ui-components/src/components/SchemaForm/Widgets/CheckboxesWidget/utils.ts:12]: The disabled should include this condition `disabled={disabled || itemDisabled || readonly}` not only the itemDisabled

File: packages/react-ui-components/src/components/SchemaForm/Widgets/CheckboxesWidget/utils.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	}
+
+	const allButtonsSelected = selectedOptions.length === allOptions.length;
+
+	// check if the user clicked on 'All' button
+	if (allButton) {
+		if (selectedOptionValue === ALL_BUTTON_VALUE) {
+			// reset to the first option selected
+			if (allButtonsSelected) {
+				return [allOptions[0].value];
```
Comment (ricardogoncalves89) [inline packages/react-ui-components/src/components/SchemaForm/Widgets/CheckboxesWidget/utils.ts:65]: I think you should only force this selection if the field is required.

File: packages/react-ui-components/src/components/SchemaForm/Widgets/CheckboxesWidget/types.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+export interface ICheckboxConfig {
+	multiple: boolean;
+	allButton: boolean;
+	minItems: number;
+	maxItems: number;
+}
```
Comment (ricardogoncalves89) [inline packages/react-ui-components/src/components/SchemaForm/Widgets/CheckboxesWidget/types.ts:6]: you can probably add the `readonly` and `disabled` props so you can use them inside helpers.

**Review by ricardogoncalves89 — CHANGES_REQUESTED** [review]

File: packages/react-ui-components/src/components/SchemaForm/Widgets/CheckboxesWidget/utils.ts (hunk 1 above)
Comment (djcouto) [reply packages/react-ui-components/src/components/SchemaForm/Widgets/CheckboxesWidget/utils.ts:12]: Absolutely 👍

File: packages/react-ui-components/src/components/SchemaForm/Widgets/CheckboxesWidget/utils.ts (hunk 2 above)
Comment (djcouto) [reply packages/react-ui-components/src/components/SchemaForm/Widgets/CheckboxesWidget/utils.ts:65]: This makes sense. I'll add a condition here to check `required || minItems > 0`.

File: packages/react-ui-components/src/components/SchemaForm/Widgets/CheckboxesWidget/utils.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	enumDisabled: string[],
+	{ multiple, allButton, disabled, readonly }: ICheckboxConfig
+): IRadioButton[] =>
+	options.reduce<IRadioButton[]>(
+		(accumulator, { label, value }) => [
+			...accumulator,
+			{
+				label,
+				value,
+				disabled: disabled || enumDisabled.indexOf(value) !== -1 || readonly
```
Comment (miguelpinto98) [inline packages/react-ui-components/src/components/SchemaForm/Widgets/CheckboxesWidget/utils.ts:16]: suggestion:
`enumDisabled.includes(value)`

File: packages/react-ui-components/src/components/SchemaForm/Widgets/CheckboxesWidget/utils.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+			  ]
+			: []
+	);
+
+export const buildSelectedRadioButtons = <T extends { label: string; value: string }>(
+	selectedOptions: string[],
+	allOptions: T[],
+	{ multiple, allButton }: ICheckboxConfig
+): { [key: string]: boolean } => {
+	if (multiple && allButton) {
```
Comment (miguelpinto98) [inline packages/react-ui-components/src/components/SchemaForm/Widgets/CheckboxesWidget/utils.ts:34]: can we combine both ifs?

File: packages/react-ui-components/src/components/SchemaForm/Widgets/CheckboxesWidget/utils.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+					return allOptions.slice(0, minItems).map(({ value }) => value);
+				}
+
+				return [];
+			}
+
+			return allOptions.map(({ value }) => value);
+		}
+
+		if (allButtonsSelected) {
```
Comment (miguelpinto98) [inline packages/react-ui-components/src/components/SchemaForm/Widgets/CheckboxesWidget/utils.ts:79]: Probably here we only want to check for the minimum items select to return the selected options

**Review by miguelpinto98 — CHANGES_REQUESTED** [review]
Good work supporting this functionality on the widget. I only left some suggestions.

File: packages/react-ui-components/src/components/SchemaForm/Widgets/CheckboxesWidget/utils.ts [hunk 7]
```diff
@@ -0,0 +1,100 @@
+import { IRadioButton } from '@kelvininc/ui-components';
+import { ALL_BUTTON_VALUE } from './config';
+import { ICheckboxConfig } from './types';
+
+export const buildRadioButtons = <T extends { label: string; value: string }>(
```
Comment (pedrorebelocosta) [inline packages/react-ui-components/src/components/SchemaForm/Widgets/CheckboxesWidget/utils.ts:5]: is there any technical limitation that forbids typing the object that T extends from?

**Review by pedrorebelocosta — CHANGES_REQUESTED** [review]

**Review by pedrorebelocosta — APPROVED** [review]

**Review by miguelpinto98 — APPROVED** [review]
👍

**Comment by miguelpinto98** [comment]
@ricardogoncalves89 can this PR be merged?

**Review by ricardogoncalves89 — APPROVED** [review]

**Pull request merged**